lua/ttt2/libraries/database.lua

```diff
@@ -0,0 +1,1419 @@
+
+		local hasAdminAccess = isAdmin and accessLevel <= DATABASE_ACCESS_ADMIN
+		local hasReadAccess = hasAdminAccess or accessLevel <= DATABASE_ACCESS_ANY
+
+		return hasReadAccess, hasAdminAccess
+	end
+
+	---
+	-- Synchronizes all registered Databases with the given players defined by the plyIdentifier
+	-- @note This is used internally to sync between server and client, you dont need to call it manually
+	-- @param string plyIdentifier the player identifier to determine who receives the message, defined in `SEND_TO_PLY_`-enums or can be a plyID64
+	-- @param[opt] string identifier the identifier used to get correct onreceive functions
+	-- @realm server
+	-- @internal
+	function database.SyncRegisteredDatabases(plyIdentifier, identifier)
+		local tableCount = #registeredDatabases
+
+		for databaseNumber = 1, tableCount do
+			--contains additional data in case an identifier is given
+			local dataRegister = {
+				index = databaseNumber,
+				identifier = identifier,
+				tableCount = tableCount
+			}
+			SendUpdateNextTick(MESSAGE_REGISTER, dataRegister, plyIdentifier)
+
+			local defaultData = registeredDatabases[databaseNumber].defaultData
+
+			-- Also sync registered item-specific defaults if available
+			if table.IsEmpty(defaultData) then continue end
+
+			local dataDefault = {
+				index = databaseNumber,
+				sendTable =  true,
+				defaultData = defaultData
+			}
+
+			SendUpdateNextTick(MESSAGE_GET_DEFAULTVALUE, dataDefault, plyIdentifier)
+		end
+	end
+
+	---
+	-- Call this when you want to setup a database that needs to be accessible by server and client
+	-- If you dont call this function before anything else, it wont work. Choose any name as accessName so that others can easily use it.
+	-- @param string databaseName the real name of the database
+	-- @param string accessName the name to quickly access databases and differentiate between a pseudo used accessName and the migrated actual databaseName
+	-- @param table savingKeys the savingKeys = {keyName = {typ, bits, default, ..}, ..} defining the keyNames and their information
+	-- @param[default = DATABASE_ACCESS_ADMIN] number accessLevel the access level needed to get values of a database, defined in `DATABASE_ACCESS_`-enums (_ANY, _ADMIN, _SERVER)
+	-- @note If accessLevel is set to DATABASE_ACCESS_SERVER it fully prevents any client read- and write-access, whereas DATABASE_ACCESS_ANY only gives read-, but not write-access to anyone
+	-- @param[opt] table additionalData the data that doesnt belong to a database but might be needed for other purposes like enums
+	-- @return bool isSuccessful if the database exists and is successfully registered
+	-- @realm server
+	function database.Register(databaseName, accessName, savingKeys, accessLevel, additionalData)
+		accessLevel = accessLevel or DATABASE_ACCESS_ADMIN
+
+		-- Create Sql table if not already done
+		if not sql.CreateSqlTable(databaseName, savingKeys) or not isstring(accessName) then
+			return false
+		end
```

Comment:
Well I wanted it in there, because it might be, that that table was deleted and in the f1 menu it will be only registered in case its needed. So I thought instead of making another create, I just create it in case it's not there to minimize casualties.
It also makes sure that there really is a database it can call.

Reply:
Valid point. could we place the creation of the database behind a flag instead then?

Reply:
Hm not sure if that makes sense, I started to draft it and came to a point where it seems pointless. I would still need to check if the table really exists and that is the same sql.CreateSqlTable does before it creates one. Also the user would still need to give me the databaseName and the savingKeys, so my library can handle it. So I could mention that in the documentation.

Im not sure why someone would try to register a database that doesnt exist. And if it exists its not created anyways. I dont really see the point in that flag, which freedom it would deliver. Only thing I see is, that I could throw an error if the user says it exists but it doesnt.